configuration_broker/config_broker/config_broker.cc

```diff
@@ -0,0 +1,343 @@
+// add a new item and search for an item (there is no concept of
+// deleting an item). The total number of items is limited by the
+// the set of static sealed capabilities.
+//
+InternalConfigitem *configData = nullptr;
+
+//
+// Sealing Types
+//
+#define CONFIG_WRITE STATIC_SEALING_TYPE(WriteConfigKey)
+#define CONFIG_READ STATIC_SEALING_TYPE(ReadConfigKey)
+#define CONFIG_PARSER STATIC_SEALING_TYPE(ParserConfigKey)
+
+//
+// Unseal a config capability.
+//
+ConfigToken *config_capability_unseal(SObj sealedCap, SKey key)
+{
+	ConfigToken *token = token_unseal(key, Sealed<ConfigToken>{sealedCap});
+
+	if (token == nullptr)
+	{
+		Debug::log("invalid config capability {}", sealedCap);
+		return nullptr;
+	}
+
+	Debug::log("Unsealed id: {} size: {} interval: {} item: {}",
+	           token->id,
+	           token->size,
+	           token->updateInterval,
+	           token->Name);
+
+	if (token->id == 0)
+	{
+		static uint16_t nextId = 1;
+
+		// Assign an ID so we can track each capability if needed
+		token->id = nextId++;
+	}
+
+	return token;
+}
+
+//
+// Find a Config by name.  If it doesn't already exist
+// create one.  Use a LockGuard to protect against two
+// threads trying to create the same item.
+//
+FlagLock lockFindOrCreate;
+
+InternalConfigitem *find_or_create_config(ConfigToken *token)
+{
+	LockGuard g{lockFindOrCreate};
+
+	for (auto c = configData; c != nullptr; c = c->next)
+	{
+		if (strcmp(c->name, token->Name) == 0)
+		{
+			return c;
+		}
+	}
+
+	// Allocate a Config object
+	InternalConfigitem *c =
+	  static_cast<InternalConfigitem *>(malloc(sizeof(InternalConfigitem)));
+
+	if (c != nullptr)
+	{
+		// Use the Name from the token that triggered the creation
+		// as the name value, since sealed objects are guaranteed not
+		// to be deallocated.
+		c->name    = token->Name;
+		c->version = 0;
+		c->data    = nullptr;
+
+		// Insert it at the start of the list
+		c->next    = configData;
+		configData = c;
+	}
+
+	return c;
+};
+
+//
+// Set a new value for the configuration item described by
+// the capability.
+//
+int __cheri_compartment("config_broker")
+  set_config(SObj sealedCap, const char *json)
+{
+	Debug::log("thread {} Set config called with {} {}",
+	           thread_id_get(),
+	           sealedCap,
+	           json);
+
+	//
+	// Check that we've been given a valid capability
+	//
+	ConfigToken *token = config_capability_unseal(sealedCap, CONFIG_WRITE);
+	if (token == nullptr)
+	{
+		Debug::log("Invalid set config capability: {}", sealedCap);
+		return -EPERM;
+	}
+
+	//
+	// We need to load all the parsers before we can
+	// accept any updates. Rather than have a thread in
+	// the parser compartment that runs to do that and
+	// then has to hang around (as threads can't exit) we
+	// use the first thread that tries to load a value to
+	// make a cross-comparement call to parserInit()
+	//
+	static bool ParsersLoaded = false;
+
+	if (!ParsersLoaded)
+	{
+		parserInit();
+		ParsersLoaded = true;
+	}
+
```

Comment:
I'm guessing because parserInit is a cross compartment call if I make it static then I get a linker error ?

```
error: ld.lld: error: undefined symbol: __library_export_libcalls___cxa_guard_acquire
>>> referenced by __cpp_config_broker.cc.cc
>>>               build/cheriot/cheriot/release/config_broker.compartment:()

ld.lld: error: undefined symbol: __library_export_libcalls___cxa_guard_release
>>> referenced by __cpp_config_broker.cc.cc
>>>               build/cheriot/cheriot/release/config_broker.compartment:()
```